src/C-interface/distributed2d/bml_diagonalize_distributed2d_typed.c

```diff
@@ -288,11 +298,195 @@ void TYPED_FUNC(
+        // convert local matrix to dense
+        amat = bml_convert(Alocal, dense, A->matrix_precision,
+                           -1, sequential);
+        zmat = bml_convert(eigenvectors->matrix, dense, A->matrix_precision,
+                           -1, sequential);
+    }
+
+    // transpose to satisfy column major ELPA convention
+    // (global matrix assumed symmetric, so no need for communications)
+    if (A->myprow != A->mypcol)
+        bml_transpose(amat);
+
+    REAL_T *z = bml_get_data_ptr(zmat);
+    assert(z != NULL);
+    REAL_T *a = bml_get_data_ptr(amat);
+    assert(a != NULL);
+
+    /* Solve EV problem */
+    // interface: see elpa_generic.h
+    // handle  handle of the ELPA object, which defines the problem
+    // a       device pointer to matrix a in GPU memory
+    // ev      on return: pointer to eigenvalues in GPU memory
+    // q       on return: pointer to eigenvectors in GPU memory
+    // error   on return the error code, which can be queried with elpa_strerr()
+    LOG_DEBUG("Call ELPA eigensolver");
+#if defined(SINGLE_REAL) || defined(SINGLE_COMPLEX)
+    float *ev;
+    magma_int_t ret = magma_smalloc(&ev, na);
+#else
+    double *ev;
+    magma_int_t ret = magma_dmalloc(&ev, na);
+#endif
+    assert(ret == MAGMA_SUCCESS);
+#if defined(SINGLE_REAL)
+    elpa_eigenvectors_float(handle, a, ev, z, &error_elpa);
+#endif
+#if defined(DOUBLE_REAL)
+    elpa_eigenvectors_double(handle, a, ev, z, &error_elpa);
+#endif
+#if defined(SINGLE_COMPLEX)
+    elpa_eigenvectors_float_complex(handle, a, ev, z, &error_elpa);
+#endif
+#if defined(DOUBLE_COMPLEX)
+    elpa_eigenvectors_double_complex(handle, a, ev, z, &error_elpa);
+#endif
+
+    assert(error_elpa == ELPA_OK);
+    // copy eigenvalues to CPU
+    LOG_DEBUG("copy eigenvalues to CPU");
+#if defined(SINGLE_REAL) || defined(SINGLE_COMPLEX)
+    float *tmp = malloc(na * sizeof(float));
+    magma_sgetvector(na, ev, 1, tmp, 1, bml_queue());
+#endif
+#if defined(DOUBLE_REAL) || defined(DOUBLE_COMPLEX)
+    double *tmp = malloc(na * sizeof(double));
+    magma_dgetvector(na, ev, 1, tmp, 1, bml_queue());
+#endif
+    magma_queue_sync(bml_queue());
+
+    REAL_T *ev_ptr = eigenvalues;
+    for (int i = 0; i < A->N; i++)
+        ev_ptr[i] = (REAL_T) tmp[i];
+    free(tmp);
+
+    magma_free(ev);
+
+    bml_deallocate(&amat);
+    if (bml_get_type(Alocal) != dense)
+    {
+        bml_deallocate(&(eigenvectors->matrix));
+        eigenvectors->matrix =
+            bml_convert(zmat, bml_get_type(Alocal), A->matrix_precision,
+                        A->M / A->npcols, sequential);
+        bml_deallocate(&zmat);
+    }
+
+    elpa_deallocate(handle, &error_elpa);
+}
+#endif
+
+void TYPED_FUNC(
+    bml_diagonalize_distributed2d) (
+    bml_matrix_distributed2d_t * A,
+    void *eigenvalues,
+    bml_matrix_distributed2d_t * eigenvectors)
+{
+#ifdef BML_USE_ELPA
+    TYPED_FUNC(bml_diagonalize_distributed2d_elpa) (A, eigenvalues,
```

Comment:
Function implementing interface to ELPA is called here

Reply:
Oh right. Thanks for pointing me in the right direction 😄